Correctly detect assets when fee is nominated in received asset

yabirgb opened the pull request.

LefterisJP commented on rotkehlchen/tests/unit/test_accounting.py:
```diff
@@ -918,6 +921,52 @@ def test_fees_count_in_cost_basis(accountant):
+    warnings = accountant.msg_aggregator.consume_warnings()
+    assert len(warnings) == 0
+    errors = accountant.msg_aggregator.consume_errors()
+    assert len(errors) == 0
```
please also assert here for the general P/L calculation to make sure it is correct.

LefterisJP commented on rotkehlchen/accounting/events.py:
```diff
@@ -247,15 +247,28 @@ def add_buy_and_corresponding_sell(
         """
+        # In the case where the fee asset is nominated in the asset that we receive
+        # take care that the fee is processed once we have the received asset in possesion.
+        if bought_asset == fee_currency:
```
I really don't understand what you are doing here or why. Same in `add_sell_and_corresponding_buy`

yabirgb replied:
What this does is to ensure that if the fee asset is the same as the asset received the fee is processed once we have the asset. So what it does is to choose where to put the fee, if in the sell part or the buy part of the trade

LefterisJP commented on rotkehlchen/tests/unit/test_accounting.py:
```diff
@@ -541,7 +542,8 @@ def test_settlement_buy(accountant):
         history_list=history,
     )
-    assert accountant.events.cost_basis.get_calculated_asset_amount('BTC').is_close('3.9908725')
+    # asset amount here is 3.9908725 + 0.0012 (from the fee)
```
How can that be possible? The fee is something you pay. If anything the `0.0012` should be deducted from somewhere.

yabirgb replied:
Yep. My fault, there is something still off. Going to check it

LefterisJP commented on rotkehlchen/tests/unit/test_accounting.py:
```diff
@@ -598,7 +600,8 @@ def test_margin_events_affect_gained_lost_amount(accountant):
         margin_list=margin_history,
     )
-    assert accountant.events.cost_basis.get_calculated_asset_amount('BTC').is_close('3.7468')
+    # here asset amount is 3.7468 + 0.0012
```
same as above here

LefterisJP reviewed (commented):
With a quick look here something does not look right

yabirgb commented:
So about the changes in the PR:

1. The values I modified in the tests was my mistake. I misunderstood what was being calculated in there. Reverted that changes and now the tests are green too.
2. Included a tests to check that the PnL is correctly computed in the new case we are testing
3. What the change does is the following: The issue appears when we compute the fee before acquiring the asset. The logic I introduced checks if the asset of the Fee is nominated in the asset we obtain. The accounting splits the trade in a buy and a sell and what I do is to include the fee once the asset has been bought if it falls in the special case.
4. I added a test to verify that it works

LefterisJP commented on rotkehlchen/accounting/events.py:
```diff
@@ -247,15 +247,28 @@ def add_buy_and_corresponding_sell(
         """
+        # In the case where the fee asset is nominated in the asset that we receive
+        # take care that the fee is processed once we have the received asset in possesion.
+        if bought_asset != fee_currency:
```
Hmm but if what you say is the case shouldn't it be:

`if bought_asset == fee_currency:`

So when the fee asset is the same as the asset we receive, we push the fee to the sell?

yabirgb replied:
Post mortem of my post mortem: This logic here is not needed since what you do first is to create the buy object. The else statement here was making sure that the fee was procesed in the buy part, so you were getting the buy and right after they fee substraction but for how the accounting is processed it doesn't matter for this case the where to put the fee since is always a buy before sell. 

So to keep it short: This `!=` was there to make sure that the fee was just after the acquisition but after reading it again is not needed because is always processed correctly. Just removed this logic as is not necessary. It is needed in the `add_sell_and_corresponding_buy`

LefterisJP commented on rotkehlchen/accounting/events.py:
```diff
@@ -458,6 +458,19 @@ def add_sell_and_corresponding_buy(
         - RemoteError if there is a problem reaching the price oracle server
         or with reading the response returned by the server
         """
+        # In the case where the fee asset is nominated in the asset that we receive
+        # take care that the fee is processed once we have the received asset in possesion.
+        if selling_asset == fee_currency:
```
Again here the comment says a different thing than what the code does.

Comment: 
> In the case where the fee asset is nominated in the asset that we receive

Code: 
`if selling_asset == fee_currency`

For them to be equal it should be:
`if receiving_asset == fee_currency:`

Also reading the if/else I am confused.

If selling_asset == fee_currency -> the same thing as before happens
otherwise -> you move the fee asset/amount to the buy instead of the sell.


What I assume you wanted to do and would also match the comment would be:

```python
if receiving_asset == fee_currency:
     buy_fee_asset = fee_currency
     buy_fee_amount = fee_amount
     sell_fee_asset = None
     sell_fee_amount = Fee(ZERO)
else:
      buy_fee_asset = None
      buy_fee_amount: Optional[Fee] = Fee(ZERO)
      sell_fee_asset = fee_currency
      sell_fee_amount = fee_amount
```

LefterisJP reviewed (commented):
<img src="https://media4.giphy.com/media/8nmvR3jAxnl2o/giphy.gif"/>

LefterisJP reviewed (approved):
all green. LGTM.
